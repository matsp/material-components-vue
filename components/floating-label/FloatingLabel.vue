<template>
  <label
    :class="classes"
    class="mdc-floating-label"
    v-bind="$attrs"
    @_init="onParentInit"
  >
    <slot />
  </label>
</template>

<script>
import { MDCFloatingLabel } from '@material/floating-label'

import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  inject: {
    getLabel: {
      default: null
    }
  },
  props: {
    floatAbove: {
      type: Boolean,
      default: false
    },
    shake: {
      type: Boolean,
      default: false
    }
  },
  data () {
    return {
      mdcFloatingLabel: undefined
    }
  },
  computed: {
    classes () {
      return {
        'mdc-floating-label--float-above': this.floatAbove,
        'mdc-floating-label--shake': this.shake
      }
    }
  },
  mounted () {
    if (!(this.getLabel instanceof Function)) { // can not be init by parent
      this.mdcFloatingLabel = MDCFloatingLabel.attachTo(this.$el)
    }
  },
  beforeDestroy () {
    if (this.mdcFloatingLabel instanceof MDCFloatingLabel) {
      this.mdcFloatingLabel.destroy()
    }
  },
  methods: {
    onParentInit () {
      const label = this.getLabel()
      if (label instanceof MDCFloatingLabel) {
        if (this.mdcFloatingLabel instanceof MDCFloatingLabel) this.mdcFloatingLabel.destroy()
        this.mdcFloatingLabel = label
      }
    }
  }
}
</script>
