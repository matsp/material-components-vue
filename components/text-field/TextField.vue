<template>
  <div
    :class="classes"
    class="mdc-text-field"
  >
    <slot name="leadingIcon" />
    <slot
      v-if="textarea"
      name="characterCounter"
    />
    <input
      v-if="!textarea"
      :value="value"
      class="mdc-text-field__input"
      v-bind="$attrs"
      @input="$emit('model', $event.target.value)"
      v-on="$listeners"
    >
    <textarea
      v-if="textarea"
      :value="value"
      class="mdc-text-field__input"
      v-bind="$attrs"
      @input="$emit('model', $event.target.value)"
      v-on="$listeners"
    />
    <div
      v-if="textarea || outlined"
      class="mdc-notched-outline"
    >
      <div class="mdc-notched-outline__leading" />
      <div
        v-if="$slots.default"
        class="mdc-notched-outline__notch"
      >
        <slot />
      </div>
      <div class="mdc-notched-outline__trailing" />
    </div>
    <slot v-if="$slots.default && !textarea && !outlined" />
    <slot name="trailingIcon" />
    <slot
      v-if="!outlined"
      name="bottomLine"
    />
  </div>
</template>

<script>
import { MDCTextField } from '@material/textfield'

import { baseComponentMixin, themeClassMixin } from '../base'
import { MDCComponent } from '@material/base/component'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  model: {
    prop: 'value',
    event: 'model'
  },
  provide () {
    return {
      getLabel: this.getLabel,
      getLineRipple: this.getLineRipple,
      getOutline: this.getOutline,
      getHelperText: this.getHelperText,
      getCharacterCounter: this.getCharacterCounter,
      getLeadingIcon: this.getLeadingIcon,
      getTrailingIcon: this.getTrailingIcon
    }
  },
  props: {
    value: {
      type: String,
      default: ''
    },
    disabled: {
      type: Boolean,
      default: false
    },
    fullWidth: {
      type: Boolean,
      default: false
    },
    outlined: {
      type: Boolean,
      default: false
    },
    dense: {
      type: Boolean,
      default: false
    },
    focused: {
      type: Boolean,
      default: false
    },
    textarea: {
      type: Boolean,
      default: false
    },
    useNativeValidation: {
      type: Boolean,
      default: true
    },
    valid: {
      type: Boolean,
      default: true
    }
  },
  data () {
    return {
      mdcTextField: undefined,
      slotObserver: undefined,
      noLabel: false,
      hasLeadingIcon: false,
      hasTrailingIcon: false
    }
  },
  computed: {
    classes () {
      return {
        'mdc-text-field--fullwidth': this.fullWidth,
        'mdc-text-field--with-leading-icon': this.hasLeadingIcon,
        'mdc-text-field--with-trailing-icon': this.hasTrailingIcon,
        'mdc-text-field--outlined': this.outlined,
        'mdc-text-field--dense': this.dense,
        'mdc-text-field--focused': this.focused, // won't change the actual activeElement
        'mdc-text-field--textarea': this.textarea,
        'mdc-text-field--no-label': this.noLabel
      }
    }
  },
  watch: {
    useNativeValidation () {
      this.mdcTextField.useNativeValidation = this.useNativeValidation
    },
    valid () {
      this.mdcTextField.valid = this.valid
    },
    value () {
      this.mdcTextField.value = this.value
    },
    disabled () {
      this.mdcTextField.disabled = this.disabled
    },
    classes () {
      this.$nextTick(() => this.reInstantiate())
    }
  },
  mounted () {
    this.updateSlot()
    this.slotObserver = new MutationObserver(() => this.updateSlot())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })
    this.instantiate()
  },
  beforeDestroy () {
    this.mdcTextField.destroy()
    this.slotObserver.disconnect()
  },
  methods: {
    updateSlot () {
      this.noLabel = this.$el.querySelector('.mdc-floating-label') == null
      this.hasLeadingIcon = this.$slots.leadingIcon != null
      this.hasTrailingIcon = this.$slots.trailingIcon != null

      // to make our icons compatible with version 0.x.y
      if (this.hasLeadingIcon) {
        this.$slots.leadingIcon.forEach(n => {
          if (n.elm instanceof Element) {
            n.elm.classList.add('mdc-text-field__icon')
          }
        })
      }
      if (this.hasTrailingIcon) {
        this.$slots.trailingIcon.forEach(n => {
          if (n.elm instanceof Element) {
            n.elm.classList.add('mdc-text-field__icon')
          }
        })
      }

      this.checkConfig()
    },
    reInstantiate () {
      this.mdcTextField.destroy()
      this.instantiate()
    },
    instantiate () {
      this.mdcTextField = MDCTextField.attachTo(this.$el)
      this.mdcTextField.useNativeValidation = this.useNativeValidation
      this.mdcTextField.valid = this.valid
      this.mdcTextField.disabled = this.disabled
      this.$nextTick(() => { // wait for the DOM change
        // tell all the children that the parent is initialized
        if (this.mdcTextField.label_ instanceof MDCComponent) {
          this.mdcTextField.label_.emit('_init')
        }
        if (this.mdcTextField.outline_ instanceof MDCComponent) {
          this.mdcTextField.outline_.emit('_init')
        }
        if (this.mdcTextField.lineRipple_ instanceof MDCComponent) {
          this.mdcTextField.lineRipple_.emit('_init')
        }
        if (this.mdcTextField.helperText_ instanceof MDCComponent) {
          this.mdcTextField.helperText_.emit('_init', this.mdcTextField.helperText_)
        }
        if (this.mdcTextField.characterCounter_ instanceof MDCComponent) {
          this.mdcTextField.characterCounter_.emit('_init', this.mdcTextField.characterCounter_)
        }
        if (this.mdcTextField.leadingIcon_ instanceof MDCComponent) {
          this.mdcTextField.leadingIcon_.emit('_init')
        }
        if (this.mdcTextField.trailingIcon_ instanceof MDCComponent) {
          this.mdcTextField.trailingIcon_.emit('_init')
        }
      })
    },
    checkConfig () {
      if (this.fullWidth && !this.noLabel && !this.textarea) {
        console.warn(
          'Do not use floating label with a full width text input. ' +
          'See https://github.com/material-components/material-components-web/tree/master/packages/mdc-textfield#full-width'
        )
      }

      if (this.fullWidth && this.outlined && !this.textarea) {
        console.warn(
          'Do not use outlined style on full width text input. ' +
          'See: https://github.com/material-components/material-components-web/tree/master/packages/mdc-textfield#full-width'
        )
      }
    },
    getLabel () {
      return this.mdcTextField.label_
    },
    getLineRipple () {
      return this.mdcTextField.lineRipple_
    },
    getOutline () {
      return this.mdcTextField.outline_
    },
    getHelperText () {
      return this.mdcTextField.helperText_
    },
    getCharacterCounter () {
      return this.mdcTextField.characterCounter_
    },
    getLeadingIcon () {
      return this.mdcTextField.leadingIcon_
    },
    getTrailingIcon () {
      return this.mdcTextField.trailingIcon_
    }
  }
}
</script>
