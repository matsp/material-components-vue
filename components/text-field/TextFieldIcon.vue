<template>
  <i
    class="mdc-text-field__icon"
    :class="classes"
    v-bind="attrs"
    v-on="$listeners"
    @_init="onParentInit"
  >
    <slot>{{ icon }}</slot>
  </i>
</template>

<script>
import { MDCTextFieldIcon } from '@material/textfield/icon'

export default {
  inject: {
    getLeadingIcon: {
      default: null
    },
    getTrailingIcon: {
      default: null
    }
  },
  props: {
    clickable: {
      type: Boolean,
      default: true
    },
    icon: {
      type: String,
      default: ''
    }
  },
  data () {
    return {
      attrs: Object.assign({}, this.$attrs),
      mdcTextFieldIcon: undefined
    }
  },
  computed: {
    classes () {
      return {
        'material-icons': this.icon.length > 0
      }
    }
  },
  watch: {
    clickable (val) {
      if (val) {
        this.$set(this.attrs, 'tabindex', '0')
        this.$set(this.attrs, 'role', 'button')
      } else {
        this.$delete(this.attrs, 'tabindex')
        this.$delete(this.attrs, 'role')
      }
    }
  },
  mounted () {
    if (this.clickable) {
      this.$set(this.attrs, 'tabindex', '0')
      this.$set(this.attrs, 'role', 'button')
    }
    if (!(this.getLeadingIcon instanceof Function || this.getTrailingIcon instanceof Function)) { // can not be init by parent
      this.mdcTextFieldIcon = MDCTextFieldIcon.attachTo(this.$el)
    }
  },
  beforeDestroy () {
    if (this.mdcTextFieldIcon instanceof MDCTextFieldIcon) {
      this.mdcTextFieldIcon.destroy()
    }
  },
  methods: {
    onParentInit () {
      const leadingIcon = this.getLeadingIcon()
      const trailingIcon = this.getTrailingIcon()
      const mdcTextFieldIcon = leadingIcon instanceof MDCTextFieldIcon && leadingIcon.root_ === this.$el ? leadingIcon : trailingIcon
      if (this.mdcTextFieldIcon instanceof MDCTextFieldIcon) this.mdcTextFieldIcon.destroy()
      this.mdcTextFieldIcon = mdcTextFieldIcon
    }
  }
}
</script>
