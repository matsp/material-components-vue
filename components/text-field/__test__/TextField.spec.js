import 'mutationobserver-shim'
import { mount } from '@vue/test-utils'
import TextField from '../TextField.vue'
import TextFieldIcon from '../TextFieldIcon.vue'
import FloatingLabel from '../../floating-label/FloatingLabel.vue'

describe('Text Field', () => {
  it('should mount', () => {
    const wrapper = mount(TextField)
    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.vm.$data.mdcTextField).toBeDefined()
  })

  it('should render with no prop', () => {
    const wrapper = mount(TextField)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-text-field')
    expect(wrapper.classes()).toContain('mdc-text-field--no-label')
    expect(wrapper.find('input').attributes('disabled')).toBeUndefined()
  })

  it('should render as focused', () => {
    const wrapper = mount(TextField)
    const input = wrapper.find('input')
    input.trigger('focus')
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-text-field--focused')
  })

  it('should render as disabled', () => {
    const wrapper = mount(TextField, {
      propsData: {
        disabled: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-text-field--disabled')
    expect(wrapper.find('input').attributes('disabled')).toBeDefined()
  })

  it('should render as outlined', () => {
    const wrapper = mount(TextField, {
      propsData: {
        outlined: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-text-field--outlined')
    expect(wrapper.find('.mdc-notched-outline').exists()).toBe(true)
  })

  it('should render as dense', () => {
    const wrapper = mount(TextField, {
      propsData: {
        dense: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-text-field--dense')
  })

  it('should render as fullWidth', () => {
    const wrapper = mount(TextField, {
      propsData: {
        fullWidth: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-text-field--fullwidth')
  })

  it('should warn of fullWidth text input with floating label', () => {
    const warnfn = console.warn
    console.warn = jest.fn()
    mount(TextField, {
      propsData: {
        fullWidth: true
      },
      slots: {
        default: '<m-floating-label>My Label</m-floating-label>'
      },
      stubs: {
        'm-floating-label': FloatingLabel
      }
    })
    expect(console.warn).toHaveBeenCalledWith(expect.stringContaining('floating label'))
    console.warn = warnfn
  })

  it('should warn of fullWidth text input with outlined style', () => {
    const warnfn = console.warn
    console.warn = jest.fn()
    mount(TextField, {
      propsData: {
        fullWidth: true,
        outlined: true
      }
    })
    expect(console.warn).toHaveBeenCalledWith(expect.stringContaining('outlined style'))
    console.warn = warnfn
  })

  it('should render and emit', () => {
    const wrapper = mount(TextField, {
      propsData: {
        value: 'val'
      }
    })

    const input = wrapper.find('input')
    input.setValue('test')
    expect(wrapper.emitted().model.length).toBe(1)
    expect(wrapper.emitted().model[0]).toEqual(['test'])
  })

  it('should render with leading icon', () => {
    const wrapper = mount(TextField, {
      slots: {
        leadingIcon: TextFieldIcon
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-text-field--with-leading-icon')
  })

  it('should render with trailing icon', () => {
    const wrapper = mount(TextField, {
      slots: {
        trailingIcon: TextFieldIcon
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-text-field--with-trailing-icon')
  })

  it('should render textarea', () => {
    const wrapper = mount(TextField, {
      propsData: {
        textarea: true
      }
    })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-text-field--textarea')
  })

  it('should render full-width textarea with outlined label', () => {
    const wrapper = mount(TextField, {
      propsData: {
        textarea: true,
        fullWidth: true
      },
      slots: {
        default: '<m-floating-label>My Label</m-floating-label>'
      },
      stubs: {
        'm-floating-label': FloatingLabel
      }
    })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-text-field--fullwidth')
    expect(wrapper.classes()).toContain('mdc-text-field--textarea')
    expect(wrapper.find('.mdc-notched-outline').exists()).toBe(true)
  })
})
