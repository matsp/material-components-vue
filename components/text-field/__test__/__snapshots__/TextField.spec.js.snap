// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Field should render as dense 1`] = `
<div class="mdc-text-field mdc-text-field--dense mdc-text-field--no-label">
  <!----> <input class="mdc-text-field__input">
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`Text Field should render as disabled 1`] = `
<div class="mdc-text-field mdc-text-field--no-label mdc-text-field--disabled">
  <!----> <input class="mdc-text-field__input" disabled="">
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`Text Field should render as focused 1`] = `
<div class="mdc-text-field mdc-text-field--no-label mdc-text-field--focused">
  <!----> <input class="mdc-text-field__input">
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`Text Field should render as fullWidth 1`] = `
<div class="mdc-text-field mdc-text-field--fullwidth mdc-text-field--no-label">
  <!----> <input class="mdc-text-field__input">
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`Text Field should render as outlined 1`] = `
<div class="mdc-text-field mdc-text-field--outlined mdc-text-field--no-label">
  <!----> <input class="mdc-text-field__input">
  <!---->
  <div class="mdc-notched-outline mdc-notched-outline--no-label">
    <div class="mdc-notched-outline__leading"></div>
    <!---->
    <div class="mdc-notched-outline__trailing"></div>
  </div>
  <!---->
  <!---->
</div>
`;

exports[`Text Field should render full-width textarea with outlined label 1`] = `
<div class="mdc-text-field mdc-text-field--fullwidth mdc-text-field--textarea">
  <!----> <textarea class="mdc-text-field__input"></textarea>
  <div class="mdc-notched-outline mdc-notched-outline--upgraded">
    <div class="mdc-notched-outline__leading"></div>
    <div class="mdc-notched-outline__notch"><label class="mdc-floating-label" style="transition-duration: 0s;">My Label</label></div>
    <div class="mdc-notched-outline__trailing"></div>
  </div>
  <!---->
</div>
`;

exports[`Text Field should render textarea 1`] = `
<div class="mdc-text-field mdc-text-field--textarea mdc-text-field--no-label">
  <!----> <textarea class="mdc-text-field__input"></textarea>
  <div class="mdc-notched-outline mdc-notched-outline--no-label">
    <div class="mdc-notched-outline__leading"></div>
    <!---->
    <div class="mdc-notched-outline__trailing"></div>
  </div>
  <!---->
</div>
`;

exports[`Text Field should render with leading icon 1`] = `
<div class="mdc-text-field mdc-text-field--with-leading-icon mdc-text-field--no-label"><i class="mdc-text-field__icon" tabindex="0" role="button"></i>
  <!----> <input class="mdc-text-field__input">
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`Text Field should render with no prop 1`] = `
<div class="mdc-text-field mdc-text-field--no-label">
  <!----> <input class="mdc-text-field__input">
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`Text Field should render with trailing icon 1`] = `
<div class="mdc-text-field mdc-text-field--with-trailing-icon mdc-text-field--no-label">
  <!----> <input class="mdc-text-field__input">
  <!---->
  <!---->
  <!----> <i class="mdc-text-field__icon" tabindex="0" role="button"></i> </div>
`;
